pkg/clustermesh/operator/service_exports.go

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package operator
+
+import (
+	"maps"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	mcsapitypes "github.com/cilium/cilium/pkg/clustermesh/mcsapi/types"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+type GlobalServiceExport struct {
+	ClusterServiceExports map[string]*mcsapitypes.MCSAPIServiceSpec
+}
+
+func newGlobalServiceExport() *GlobalServiceExport {
+	return &GlobalServiceExport{
+		ClusterServiceExports: map[string]*mcsapitypes.MCSAPIServiceSpec{},
+	}
+}
+
+type GlobalServiceExportCache struct {
+	log         logrus.FieldLogger
+	mutex       lock.RWMutex
+	byName      map[types.NamespacedName]*GlobalServiceExport
+	byNamespace map[string]sets.Set[*GlobalServiceExport]
+
+	// metricTotalGlobalServiceExports is the gauge metric for total of global service exports
+	metricTotalGlobalServiceExports metric.Gauge
+}
+
+func NewGlobalServiceExportCache(log logrus.FieldLogger, metricTotalGlobalServiceExports metric.Gauge) *GlobalServiceExportCache {
+	return &GlobalServiceExportCache{
+		log:                             log,
+		byName:                          map[types.NamespacedName]*GlobalServiceExport{},
+		byNamespace:                     map[string]sets.Set[*GlobalServiceExport]{},
+		metricTotalGlobalServiceExports: metricTotalGlobalServiceExports,
+	}
+}
+
+// GetServiceExports returns the service exports for a specific namespace. This function does not
+// make copy of the cluster service exports objects so those objects should not be mutated.
+func (c *GlobalServiceExportCache) GetServiceExports(namespace string) []*mcsapitypes.MCSAPIServiceSpec {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	clusterSvcExports := []*mcsapitypes.MCSAPIServiceSpec{}
+
+	if globalSvcExports, ok := c.byNamespace[namespace]; ok {
+		for globalSvcExport := range globalSvcExports {
```

Comment:
```suggestion
		for globalSvcExport := range c.byNamespace[namespace] {
```

Nit. Dropping the surrounding if, for simplicity.